coverage: Specify modules to run coverage for

stephenfin opened the pull request:
### Feature or Bugfix

Feature

### Purpose

Currently, the coverage builder lets you check for partially documented modules, but there is no mechanism to identify totally undocumented modules. Resolve this by introducing a new `coverage_modules` config option. This is a list of modules that should be documented somewhere within the documentation tree. Any modules that are specified in the configuration value but are not documented anywhere will result in a warning. Likewise, any modules that are not in the config option but are documented somewhere will result in a warning.

> **Note**
> I have a WIP series locally that will complete the documentation for this module and close out #1288, but I'm keeping that out of this to avoid making it any more complicated than necessary.

### Detail

(none)

### Relates

(none)

stephenfin committed:
coverage: Specify modules to run coverage for

Currently, the coverage builder lets you check for partially documented
modules, but there is no mechanism to identify totally undocumented
modules. Resolve this by introducing a new 'coverage_modules' config
option. This is a list of modules that should be documented somewhere
within the documentation tree. Any modules that are specified in the
configuration value but are not documented anywhere will result in a
warning. Likewise, any modules that are not in the config option but are
documented somewhere will result in a warning.

Signed-off-by: Stephen Finucane <stephen@that.guru>

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -65,6 +66,45 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
+        if sub_mod_info.name == '__main__':
+            continue
+
+        # TODO: Do we want to skip private modules (note: different from private
```
I think we can skip them by using a regex instead so let's pick them by default.

stephenfin replied:
wfm

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -65,6 +66,45 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
+        if sub_mod_info.ispkg:
+            modules |= load_modules(f'{mod_name}.{sub_mod_info.name}', ignored_module_exps)
+        else:
+            modules.add(f'{mod_name}.{sub_mod_info.name}')
```
Don't you want to exclude a submodule if the latter matches a regexpr?

stephenfin replied:
Good point. Done.

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -90,6 +130,7 @@ def init(self) -> None:
         for (name, exps) in self.config.coverage_ignore_c_items.items():
             self.c_ignorexps[name] = compile_regex_list('coverage_ignore_c_items',
                                                         exps)
+        self.modules = self.config.coverage_modules
```
I think module_names would be better.

stephenfin replied:
Agreed. Done.

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -167,11 +208,43 @@ def ignore_pyobj(self, full_name: str) -> bool:
+                try:
+                    modules |= load_modules(mod_name, self.mod_ignorexps)
+                except ImportError as err:
+                    logger.warning(__('module %s could not be imported: %s'), mod_name, err)
```
Since people like to suppress warnings, could you also decide for a type/subtype for the log record and add it to the existing ones.

stephenfin replied:
My one disagreement: none of the other `logger.warning` calls in here do this, so I think we want to do it for all or none. Can I do this as a follow-up? I've added a TODO.

picnixz replied:
Sure! Overall, we need to add those type/subtype more or less everywhere where warnings could be annoying because people request the ability to suppress them (and to turn them as non-errors), so you could have a separate issue for warnings regarding coverage-related things.

picnixz reviewed (commented):
Do we need some extra tests?

stephenfin committed:
tests: Add tests for new coverage functionality

We migrate the code we're measuring coverage for to a package so we can
validate the new module coverage functionality.

Signed-off-by: Stephen Finucane <stephen@that.guru>

stephenfin committed:
coverage: Sort modules

This prevent the table changing every time we generate the report.

Signed-off-by: Stephen Finucane <stephen@that.guru>

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -65,6 +66,44 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
     yield _add_line(col_widths, separator)
 
 
+def load_modules(mod_name: str, ignored_module_exps: list[re.Pattern[str]]) -> set[str]:
```
Actually, do you want the function to be part of the public API? or could we rename it `_load_modules`?

stephenfin replied:
Fair point. Done.

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -282,19 +354,23 @@ def _write_py_statistics(self, op: TextIO) -> None:
-            f'{len(all_objects) - len(all_documented_objects)}',
-        ])
+
+        if len(all_objects):
```
```suggestion
        if all_objects:
```

stephenfin replied:
I was pretty sure this was done for a reason. Something like this is what had been done elsewhere. I can't remember now though so...

picnixz commented on sphinx/ext/coverage.py:
```diff
@@ -271,8 +344,7 @@ def build_py_coverage(self) -> None:
         """ Outputs the table of ``op``."""
-        all_modules = set(self.py_documented.keys()).union(
-            set(self.py_undocumented.keys()))
+        all_modules = set(self.py_documented.keys()).union(set(self.py_undocumented.keys()))
```
```suggestion
        all_modules = frozenset(self.py_documented.keys() | self.py_undocumented.keys())
```

picnixz commented on tests/roots/test-ext-coverage/grog/coverage_missing.py:
```diff
@@ -0,0 +1,8 @@
+
+    def missing_a(self):
+        """An undocumented method."""
+        pass
```
```suggestion
```

picnixz commented on tests/test_ext_coverage.py:
```diff
@@ -10,8 +10,10 @@ def test_build(app, status, warning):
     app.builder.build_all()
 
     py_undoc = (app.outdir / 'python.txt').read_text(encoding='utf8')
-    assert py_undoc.startswith('Undocumented Python objects\n'
```
Any reason for this change?

stephenfin replied:
I think this is an artefact from an earlier revision of the patch. It's unnecessary but also harmless. I'll take it out if I need to respin.

picnixz reviewed (commented):
Just so that you know, I didn't review the doc content.

AA-Turner committed:
coverage: Extract _determine_py_coverage_modules into a function

AA-Turner commented on sphinx/ext/coverage.py:
```diff
@@ -65,6 +66,86 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
+
+    # if there are additional modules then we warn (but still scan)
+    additional_modules = set(seen_modules) - modules
+    if additional_modules:
```
I think this warning only captures half of what you noted -- modules that are in ``coverage_modules`` but not in ``seen_modules`` don't seem to get a warning.

```suggestion
    # if there are additional modules then we warn (but still scan)
    if additional_modules := set(seen_modules) - modules:
```

stephenfin replied:
They weren't getting a warning but they were getting captured in the report. I've added the warning nonetheless.

AA-Turner commented on sphinx/ext/coverage.py:
```diff
@@ -65,6 +66,86 @@ def _add_row(col_widths: list[int], columns: list[str], separator: str) -> Itera
     yield _add_line(col_widths, separator)
 
 
+def _load_modules(mod_name: str, ignored_module_exps: list[re.Pattern[str]]) -> set[str]:
```
I wonder if we'd be able to use ``importlib`` (e.g. [``importlib.util.find_spec``](https://docs.python.org/dev/library/importlib.html#importlib.util.find_spec); [``importlib.util.spec_from_file_location``](https://docs.python.org/dev/library/importlib.html#importlib.util.spec_from_file_location)) to avoid importing the modules, given we only want to find their names and subpackages.

stephenfin replied:
Does it matter? We won't import modules that we don't want to document as we check the module name first, and we import all the other modules anyway via later calls to `import_module`.

AA-Turner reviewed (approved):
Looks good, two notes

A

stephenfin committed:
coverage: Warn on missing modules

Signed-off-by: Stephen Finucane <stephen@that.guru>

AA-Turner commented on doc/usage/extensions/coverage.rst:
```diff
@@ -6,15 +6,58 @@
 
 This extension features one additional builder, the :class:`CoverageBuilder`.
 
+.. todo:: Write this section.
```
Will this TODO remain?

stephenfin replied:
Not long-term. I have a separate series but GitHub doesn't let me stack PRs. I'll push once this is merged.

AA-Turner commented on doc/usage/extensions/coverage.rst:
```diff
@@ -6,15 +6,58 @@
 
-Several configuration values can be used to specify what the builder
-should check:
+   .. versionadded: 7.2
```
```suggestion
   .. versionadded:: 7.2
```

stephenfin committed:
Fix typo

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

AA-Turner committed:
Merge branch 'master' into coverage-all-modules

AA-Turner committed:
Tweaks, typing, etc

AA-Turner committed:
CHANGES
